test/e2e/storage/testsuites/multivolume.go

```diff
@@ -503,3 +527,56 @@ func TestConcurrentAccessToSingleVolume(f *framework.Framework, cs clientset.Int
 		utils.CheckReadFromPath(f, pod, *pvc.Spec.VolumeMode, path, byteLen, seed)
 	}
 }
+
+// getCurrentTopologies() goes through all Nodes and returns unique driver topologies and count of Nodes per topology
+func getCurrentTopologiesNumber(cs clientset.Interface, nodes *v1.NodeList, keys []string) ([]topology, []int, error) {
+	topos := []topology{}
+	topoCount := []int{}
+
+	// TODO: scale?
+	for _, n := range nodes.Items {
+		topo := map[string]string{}
+		for _, k := range keys {
+			v, ok := n.Labels[k]
+			if ok {
+				topo[k] = v
+			}
+		}
+
+		found := false
+		for i, existingTopo := range topos {
+			if topologyEqual(existingTopo, topo) {
+				found = true
+				topoCount[i]++
+				break
+			}
+		}
+		if !found {
+			framework.Logf("found topology %v", topo)
+			topos = append(topos, topo)
+			topoCount = append(topoCount, 1)
+		}
+	}
+	return topos, topoCount, nil
+}
+
+// ensureTopologyRequirements sets nodeSelection affinity according to given topology keys for drivers that provide them
+func ensureTopologyRequirements(nodeSelection *e2epod.NodeSelection, nodes *v1.NodeList, cs clientset.Interface, topologyKeys []string, minCount int) error {
+	topologyList, topologyCount, err := getCurrentTopologiesNumber(cs, nodes, topologyKeys)
+	if err != nil {
+		return err
+	}
+	suitableTopologies := []topology{}
+	for i, topo := range topologyList {
+		if topologyCount[i] >= minCount {
+			suitableTopologies = append(suitableTopologies, topo)
+		}
+	}
+	if len(suitableTopologies) == 0 {
+		framework.Skipf("No topology with at least %d nodes found - skipping", minCount)
```

Comment:
If a driver optionally supports topology (like vsphere), and this runs on a platform where topology is not used, are we going to end up skipping the test?

Reply:
I didn't think about this... Is there a way to detect the topology support is optional?

Reply:
On vSphere without topology support, nodes won't have topology labels. Looking at `getCurrentTopologiesNumber`, it may return some weird arrays. I added a comment above.

Reply:
And the test gets skipped, because no topology has more than 1 node.

Reply:
Nodes with the "empty" topologies get counted as if they had the same topology. I tried to re-run the tests on a cluster with no topology labels and the test run looks that same (since all the nodes were actually in the same zone). So I guess it's OK.